pallets/msa/src/lib.rs

```diff
@@ -1292,6 +1353,99 @@ impl<T: Config> Pallet<T> {
+	pub fn get_msa_keys_offchain(msa_id: MessageSourceId) -> (T::BlockNumber, Vec<T::AccountId>) {
+		let msa_keys_res = get_msa_keys::<MessageSourceId, T::AccountId, T::BlockNumber>(msa_id);
+		if let Err(e) = msa_keys_res {
+			log_err!("Error getting msa keys from offchain storage: {:?}", e);
+			return (T::BlockNumber::zero(), vec![])
+		}
+		msa_keys_res.unwrap()
+	}
+
+	fn process_events(
+		msa_id: MessageSourceId,
+		events: Vec<(T::AccountId, EventType, T::BlockNumber)>,
+	) {
+		// Lock will specifically prevent multiple offchain workers from
+		// processing the same msa events at the same time
+		let lock_status = offchain_common::lock(b"msa::", msa_id.encode().as_slice(), || {
+			for (key, event_type, block) in &events {
+				match event_type {
+					EventType::Add => {
+						let add_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Add::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), *block),
+							);
+						if let Err(e) = add_result {
+							log_err!("Error adding key to offchain storage: {:?}", e);
+						}
+					},
+					EventType::Remove => {
+						let delete_result =
+							process_msa_key_event::<MessageSourceId, T::AccountId, T::BlockNumber>(
+								MSAPublicKeyDataOperation::Remove::<
+									MessageSourceId,
+									T::AccountId,
+									T::BlockNumber,
+								>(msa_id, key.clone(), *block),
+							);
+						if let Err(e) = delete_result {
+							log_err!("Error deleting key from offchain storage: {:?}", e);
+						}
+					},
+				}
+			}
+		});
+		if let offchain_common::LockStatus::Locked = lock_status {
+			log_err!("Unable to acquire lock, skipping event processing");
+		}
+	}
+
+	fn index_event(event: Event<T>) {
+		let block_number: u32 =
+			<frame_system::Pallet<T>>::block_number().try_into().unwrap_or_default();
+		let current_event_count: u16 = <MSAEventCount<T>>::get().saturating_add(1);
+		<MSAEventCount<T>>::put(current_event_count);
+		let event_key = [
+			BLOCK_EVENT_KEY,
+			block_number.encode().as_slice(),
+			current_event_count.encode().as_slice(),
+		]
+		.concat();
+		// set the event in offchain storage
+		set_offchain_index(event_key.encode().as_slice(), event);
+
+		let count_key = [BLOCK_EVENT_COUNT_KEY, block_number.encode().as_slice()].concat();
+		// Set the latest count of event in current block
+		set_offchain_index(count_key.encode().as_slice(), current_event_count);
+	}
+
+	fn read_events(block_number: T::BlockNumber) -> Vec<Event<T>> {
+		let current_block: u32 = block_number.try_into().unwrap_or_default();
+		let count_key = [BLOCK_EVENT_COUNT_KEY, current_block.encode().as_slice()].concat();
+		let optional_event_count = get_offchain_index::<u16>(count_key.encode().as_slice());
+		let mut events = vec![];
+		let event_count =
+			if let Some(event_count) = optional_event_count { event_count } else { return events };
+
```

Comment:
checks if any event count information is indexed if so get the count and find all MSA events index by counter